### ISSUE TITLE
Category annotations on individual methods with Parameterized classes

### ISSUE THREAD

**Issue by brianlum** [junit-team/junit#761]
I'm a huge fan of the Category annotations.  I've been using so that we can use the same code to test against multiple versions of our API (I label each test with a minimum version number and the version numbers inherit from the newer versions so that the newest version will run against all tests).

I was hoping to parameterized some of our tests though, but I cannot as I get "Category annotations on Parameterized classes are not supported on individual methods."  I realize that this is currently expected.

Can we get Category annotations on individual methods with Parameterized classes?

**Comment by DiaaAttia** [comment]
I recommend to take individual classes through http://preply.com/en

**Comment by marcphilipp** [comment]
@DiaaAttia Is this GitHub issues spam?

**Comment by Tibor17** [comment]
For instance Maven does not use Categories suite to filter out tests. Instead a Filter is injected in every runner to do this job instead.
Maybe a hint with an extension point
`@RunWith(value = Parameterized.class, filters={org.junit.experimental.categories.Categories.CategoryFilter.class})`

**Comment by angaarchand** [comment]
Is the issue of running of ' method level Categories with Parameterised classes' resolved yet?

**Comment by brianlum** [comment]
yes

**Comment by angaarchand** [comment]
I'm also trying to do the same thing that is method-level Categories annotation with Parameterised class. I've tried it with Junit4.11  and Junit4.12 but it is not working for me.
@Brian : Can you please give a brief idea about how  you implemented it ?

**Comment by brianlum** [comment]
I don't have access to the code anymore, it's been a long time (this was
opened in 2013).  It worked back then.

Basically it worked using what I said in the thread:

public class APIEndpointTest {
  private client;
  public APIEndpointTest(String endpointUrl) throws Exception {
    this.client = new CustomHttpClient(endpointUrl);
  }

  @Parameterized.Parameters(name="{index}: endpoint:{0}")
  public static Collection<Object[]> getEndpoints() {
    return Arrays.asList{new String[] { "http://endpt1.com",
"http://endpoint2.com" }};
  }

  @Category(APIVersion.v2_0)
  @Test
  public void testDoSomething() throws Exception {
  }

  @Category(APIVersion.v1_9)
  @Test
  public testDoSomethingElse() throws Exception {
  }
}


And then I declared the interfaces

public interface v1_9 extends v2_0 { }
public interface v2_0 { }


The newest version is the base class, so it applies to everything.  If you
add a v2.1, then you can make v2_0 extend v2_1 and make v2_1 the base class.

Hope that makes sense.

**Comment by angaarchand** [comment]
I see. I tried a similar approach but it didn't work.  The only difference I could gauge between the structures of our code is I don't any test case in the Parameterised class. The structure is something like --

1. ChatTest.java (  file  for test class which will have the test cases which are annotated using the Junit Categories . ChatTest extends QAClientUtility).

Code Snippet :

public class ChatTest extends QAClientUtility implements XChatModuleClientEvents
{
	/**
	 * Constructor.
	 * 
	 * @param protocol
	 *            The protocol.
	 * @throws Exception
	 *             The exception.
	 */
	public ChatTest( String protocol ) throws Exception
	{
		super( protocol );
 
              // Some setup code for tests    
	}

        @Test
	@Category( OSNSRG.class )
	public void testBug21150080()
	{
		// Some test code		
	}


Note : Every such test class extends QAClientUtility. QAClientUtility extends AbsractAPIClientTest .
Inheritance structure looks like 
AbsractAPIClientTest  (Base class) -->  QAClientUtility --> ChatTest


2. QAClientUtility.java ( class having utility methods used by all the tests.  This extends AbsractAPIClientTest) 

Code Snippet :

public class QAClientUtility extends AbstractAPIClientTestCase
{
  // Variable initialisation code
   public QAClientUtility( String protocol ) throws Exception
	{
		super( protocol );
 
               //  Code for initalisation  like setting up host, test users and other things
          }

        // Utility like these

      /**
	 * This generic utility method is used to poll a element from any blocking
	 * queue list.
	 * 
	 * @param elementsList
	 * @return
	 * @throws InterruptedException
	 */
	protected <T> Object poll( BlockingQueue<T> elementsList ) throws InterruptedException
	{
		return poll( elementsList, STANDARD_POLL_TIME_60 );
	}

3. AbstractAPIClientTest.java ( every test class like ChatTest is indirectly 'extends' this class. Also, this is the parameterised class to which I'm passsing the arguments - "XML " and "JSON" as I've to run my test suite against these 2 types of Client requests.

@RunWith( Parameterized.class )
public abstract class AbstractAPIClientTestCase
{
 
       /**
	 * Public zero-arg constructor is required by JUnit.
	 */
	public AbstractAPIClientTestCase()
	{
		doSetup( true );
	}

        
        private void doSetup( boolean createNewInstance )
	{
		if ( createNewInstance )
		{
			fAPI = XAPIManager.newInstance();
		}
		else
		{
			fAPI = XAPIManager.getInstance();
		}

             // API initialisation code
         }

       	/**
	 * Constructor to be used with JUnit @Parameters. This constructor will
	 * configure XAPIManager to use either JSON or XML protocol.
	 * 
	 * @param protocol
	 *            the protocol for XAPIManager to use, either {@code XML} or
	 *            {@code JSON}.
	 */
	public AbstractAPIClientTestCase( String protocol )
	{
		this();

             // Setting properties for testing the application
             // as per values passed as arguments while running the test code

        }

	/**
	 * Return parameters passed to the constructor for each test run. Each test
	 * will be run once for each parameter array returned by this method.
	 * 
	 * @return an array of XAPIManager protocols.
	 */
	@Parameterized.Parameters
	public static Collection<Object[]> APIProtocols()
	{
		return Arrays.asList( new Object[][] { { "XML" }, { "JSON" } } );
	}

            // Some more methods code for setting properties for test code

Note that this is the Paramterized class and does not contain any test cases whatsoever.


4.  Interface for Category - SRG 

public interface OSNSRG
{

}

5. SRG implementation


import org.junit.experimental.categories.Categories;
import org.junit.experimental.categories.Categories.IncludeCategory;
import org.junit.runner.RunWith;

import waggle.client.chat.ChatTest;

@RunWith( Categories.class )
@IncludeCategory( OSNSRG.class )
@Categories.SuiteClasses( { ChatTest.class } )
public class OSNSRGTest
{

}

--------------------------------------------------------------

What I'm trying to do is here that 'run all tests marked as SRGs in the test suite' . So for that, I'm running OSNSRGTest .  The issue I'm facing is that when I annotate  'ChatTest' as SRG as a whole class then it runs fine but when I try to run the code given here (i.e. annotations on individual test case, nothing at the class level), I get the error 

"Category annotations on Parameterized classes are not supported on individual methods"

@Brian, I'm not sure if I'm doing anything (largely) different from what you posted above.  May be you can take a look and point at where I'm making the mistake. That will be of great  help. Thanks!

**Comment by stefanbirkner** [comment]
I think this issue is a duplicate of #751, which is fixed in the next JUnit release. (Please reopen the issue if I'm wrong and it is not a duplicate of #751).

**Comment by luscicli** [comment]
I tried to your build the 4.13 version but the bug persists. I found a solution for the issue.
I observed that in my test suite classes usually I didn't put the `@RunWith` annotation.
So when the runner reflected my test classes considered it like `@RunWith(Parameterized.class)` by default.
My idea was been to explicit a runner for these classes but existing default JUnit runners didn't work. 
I solved it creating a specific runner for Categorized JUnit Test Classes (derived by `org.junit.runners.ParentRunner<T>` and similar to `org.junit.runners.BlockJUnit4ClassRunner`) and now it filter my test case by categories.
Now it works for me.

I hope it was useful for us.
